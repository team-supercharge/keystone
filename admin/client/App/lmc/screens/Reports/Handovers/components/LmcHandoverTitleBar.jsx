import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { isBrowser, isTablet } from 'react-device-detect'
import { GlyphButton } from '../../../../../elemental'

export default class LmcHandoverTitleBar extends Component {
    state = {
        isClicked: false
    }

    handleClick = () => {
        this.setState(prevState => ({
            isClicked: !prevState.isClicked
        }))
        this.props.onClick()
    }

    formatTimePeriod = () => {   
        const splitAfternoon = 12; // 24hr time to split the afternoon
        const splitEvening = 17; // 24hr time to split the evening
        const hour = parseFloat(moment(this.props.createdOn).format('HH'));
        
        if (hour >= splitAfternoon && hour <= splitEvening) {
            // Between 12 PM and 5PM
            return 'Afternoon';
        } else if (hour >= splitEvening) {
            // Between 5PM and Midnight
            return 'Evening';
        }
        // Between dawn and noon
        return 'Morning';
    }

    renderCarers = () => {
        const { createdBy, witnessedBy } = this.props
        return (
            <div style={styles.carersContainer}>
                <div style={styles.carer}>
                    <span style={styles.name}>
                        By {createdBy.name.first} {createdBy.name.last}
                    </span>
                    <div 
                        className='lmc-profile-picture__handover__small'
                        style={{
                            ...styles.carerImage,
                            background: `url(${createdBy.picture || PLACEHOLDER_IMAGE})`
                        }}
                    />
                </div>
                <div style={styles.carer}>
                    <span style={styles.name}>
                        to {witnessedBy.name.first} {witnessedBy.name.last}
                    </span>
                    <div 
                        className='lmc-profile-picture__handover__small'
                        style={{
                            ...styles.carerImage,
                            background: `url(${witnessedBy.picture || PLACEHOLDER_IMAGE})`
                        }}
                    />
                </div>
            </div>
        )
    }
    
    render () { 
        const isDesktopStyles = (isBrowser || isTablet)
        return (
            <div 
                onClick={this.handleClick}
                style={styles.container}
            >
                <div style={styles.handoverTime}>
                    {`${this.formatTimePeriod()} at ${moment(this.props.createdOn).format('HH:mm')}`}
                </div>
                <div style={styles.rightContainer}>
                    { isDesktopStyles && this.renderCarers() }
                    <GlyphButton
                        className='lmc-collapse-button'
                        glyph={this.state.isClicked ? 'chevron-up' : 'chevron-down'}
                        style={styles.glyph}
                    />
                </div>
            </div>
        )
    }
}

const PLACEHOLDER_IMAGE = 'https://s3.eu-west-2.amazonaws.com/lmc-data-production/public/profile_pic_placeholder.png';

const styles = {
    carer: {
        display: 'flex',
        flexDirection: 'row',
    },
    carerImage: {
        marginRight: 10,
    },
    carersContainer: {
        display: 'flex',
        flexDirection: 'row',
    },
    container: {
        borderBottom: '1px #eaeaea solid',
        cursor: 'pointer',
        display: 'flex',
        flexDirection: 'row',
        height: 50,
        justifyContent: 'space-between',
        marginBottom: 10,
        width: '100%',
    },
    glyph: {
        paddingBottom: 10
    },
    handoverTime: {
        height: '100%',
        paddingTop: 10,
        textOverflow: 'ellipsis',
        hyphens: 'auto',
    },
    name: {
        display: 'block',
        fontSize: 12,
        opacity: 0.8,
        padding: '10px 10px 0px 0px',
    },
    rightContainer: {
        display: 'flex',
        flexDirection: 'row'
    }
}

LmcHandoverTitleBar.propTypes = {
    createdOn: PropTypes.string.isRequired,
    createdBy: PropTypes.object.isRequired,
    witnessedBy: PropTypes.object.isRequired,
    onClick: PropTypes.func.isRequired
}