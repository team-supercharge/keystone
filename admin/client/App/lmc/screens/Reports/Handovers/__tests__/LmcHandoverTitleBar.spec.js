import React from 'react'
import { shallow } from 'enzyme'
import LmcHandoverTitleBar from '../components/LmcHandoverTitleBar.jsx'

describe('LmcHandoverTitleBar', () => {
    let wrapper
    let createdOn
    let createdBy
    let witnessedBy
    let onClick = jest.fn()
    global.moment = () => ({ format: () => {} })

    beforeEach(() => {
        createdOn = new Date('1/1/2019').toString()
        createdBy = { name: { first: 'Test', last: 'Carer' }, picture: 'TestPicture' }
        witnessedBy = { name: { first: 'Test2', last: 'Carer2' }, picture: 'TestPicture2' }
        wrapper = shallow(
            <LmcHandoverTitleBar
                createdOn={createdOn}
                createdBy={createdBy}
                witnessedBy={witnessedBy}
                onClick={onClick}
            />
        )
    })

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders both carer first names', () => {
        expect(wrapper.text().includes(createdBy.name.first)).toBe(true)
        expect(wrapper.text().includes(witnessedBy.name.first)).toBe(true)
    })

    it('triggers its onClick prop when clicking its main container', () => {
        const container = wrapper.find('div').first()
        container.simulate('click')
        expect(onClick).toHaveBeenCalledTimes(1)
    })
})