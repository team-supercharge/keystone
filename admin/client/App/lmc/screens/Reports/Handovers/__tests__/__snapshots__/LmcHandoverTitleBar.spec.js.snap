// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcHandoverTitleBar renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    onClick={[Function]}
    style={
      Object {
        "borderBottom": "1px #eaeaea solid",
        "cursor": "pointer",
        "display": "flex",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "marginBottom": 10,
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": "100%",
          "hyphens": "auto",
          "paddingTop": 10,
          "textOverflow": "ellipsis",
        }
      }
    >
      Morning at undefined
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <span
            style={
              Object {
                "display": "block",
                "fontSize": 12,
                "opacity": 0.8,
                "padding": "10px 10px 0px 0px",
              }
            }
          >
            By 
            Test
             
            Carer
          </span>
          <div
            className="lmc-profile-picture__handover__small"
            style={
              Object {
                "background": "url(TestPicture)",
                "marginRight": 10,
              }
            }
          />
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <span
            style={
              Object {
                "display": "block",
                "fontSize": 12,
                "opacity": 0.8,
                "padding": "10px 10px 0px 0px",
              }
            }
          >
            to 
            Test2
             
            Carer2
          </span>
          <div
            className="lmc-profile-picture__handover__small"
            style={
              Object {
                "background": "url(TestPicture2)",
                "marginRight": 10,
              }
            }
          />
        </div>
      </div>
      <GlyphButton
        className="lmc-collapse-button"
        glyph="chevron-down"
        glyphStyle={Object {}}
        position="default"
        style={
          Object {
            "paddingBottom": 10,
          }
        }
      />
    </div>
  </div>,
  "nodes": Array [
    <div
      onClick={[Function]}
      style={
        Object {
          "borderBottom": "1px #eaeaea solid",
          "cursor": "pointer",
          "display": "flex",
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "space-between",
          "marginBottom": 10,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "height": "100%",
            "hyphens": "auto",
            "paddingTop": 10,
            "textOverflow": "ellipsis",
          }
        }
      >
        Morning at undefined
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <span
              style={
                Object {
                  "display": "block",
                  "fontSize": 12,
                  "opacity": 0.8,
                  "padding": "10px 10px 0px 0px",
                }
              }
            >
              By 
              Test
               
              Carer
            </span>
            <div
              className="lmc-profile-picture__handover__small"
              style={
                Object {
                  "background": "url(TestPicture)",
                  "marginRight": 10,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <span
              style={
                Object {
                  "display": "block",
                  "fontSize": 12,
                  "opacity": 0.8,
                  "padding": "10px 10px 0px 0px",
                }
              }
            >
              to 
              Test2
               
              Carer2
            </span>
            <div
              className="lmc-profile-picture__handover__small"
              style={
                Object {
                  "background": "url(TestPicture2)",
                  "marginRight": 10,
                }
              }
            />
          </div>
        </div>
        <GlyphButton
          className="lmc-collapse-button"
          glyph="chevron-down"
          glyphStyle={Object {}}
          position="default"
          style={
            Object {
              "paddingBottom": 10,
            }
          }
        />
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcHandoverTitleBar
        createdBy={
          Object {
            "name": Object {
              "first": "Test",
              "last": "Carer",
            },
            "picture": "TestPicture",
          }
        }
        createdOn="Tue Jan 01 2019 00:00:00 GMT+0000 (GMT)"
        onClick={[MockFunction]}
        witnessedBy={
          Object {
            "name": Object {
              "first": "Test2",
              "last": "Carer2",
            },
            "picture": "TestPicture2",
          }
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcHandoverTitleBar {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "formatTimePeriod": [Function],
        "handleClick": [Function],
        "props": Object {
          "createdBy": Object {
            "name": Object {
              "first": "Test",
              "last": "Carer",
            },
            "picture": "TestPicture",
          },
          "createdOn": "Tue Jan 01 2019 00:00:00 GMT+0000 (GMT)",
          "onClick": [MockFunction],
          "witnessedBy": Object {
            "name": Object {
              "first": "Test2",
              "last": "Carer2",
            },
            "picture": "TestPicture2",
          },
        },
        "refs": Object {},
        "renderCarers": [Function],
        "state": Object {
          "isClicked": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          onClick={[Function]}
          style={
            Object {
              "borderBottom": "1px #eaeaea solid",
              "cursor": "pointer",
              "display": "flex",
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-between",
              "marginBottom": 10,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "height": "100%",
                "hyphens": "auto",
                "paddingTop": 10,
                "textOverflow": "ellipsis",
              }
            }
          >
            Morning at undefined
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "display": "block",
                      "fontSize": 12,
                      "opacity": 0.8,
                      "padding": "10px 10px 0px 0px",
                    }
                  }
                >
                  By 
                  Test
                   
                  Carer
                </span>
                <div
                  className="lmc-profile-picture__handover__small"
                  style={
                    Object {
                      "background": "url(TestPicture)",
                      "marginRight": 10,
                    }
                  }
                />
              </div>
              <div
                style={
                  Object {
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "display": "block",
                      "fontSize": 12,
                      "opacity": 0.8,
                      "padding": "10px 10px 0px 0px",
                    }
                  }
                >
                  to 
                  Test2
                   
                  Carer2
                </span>
                <div
                  className="lmc-profile-picture__handover__small"
                  style={
                    Object {
                      "background": "url(TestPicture2)",
                      "marginRight": 10,
                    }
                  }
                />
              </div>
            </div>
            <GlyphButton
              className="lmc-collapse-button"
              glyph="chevron-down"
              glyphStyle={Object {}}
              position="default"
              style={
                Object {
                  "paddingBottom": 10,
                }
              }
            />
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          onClick={[Function]}
          style={
            Object {
              "borderBottom": "1px #eaeaea solid",
              "cursor": "pointer",
              "display": "flex",
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-between",
              "marginBottom": 10,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "height": "100%",
                "hyphens": "auto",
                "paddingTop": 10,
                "textOverflow": "ellipsis",
              }
            }
          >
            Morning at undefined
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
              }
            }
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "row",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "display": "block",
                      "fontSize": 12,
                      "opacity": 0.8,
                      "padding": "10px 10px 0px 0px",
                    }
                  }
                >
                  By 
                  Test
                   
                  Carer
                </span>
                <div
                  className="lmc-profile-picture__handover__small"
                  style={
                    Object {
                      "background": "url(TestPicture)",
                      "marginRight": 10,
                    }
                  }
                />
              </div>
              <div
                style={
                  Object {
                    "display": "flex",
                    "flexDirection": "row",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "display": "block",
                      "fontSize": 12,
                      "opacity": 0.8,
                      "padding": "10px 10px 0px 0px",
                    }
                  }
                >
                  to 
                  Test2
                   
                  Carer2
                </span>
                <div
                  className="lmc-profile-picture__handover__small"
                  style={
                    Object {
                      "background": "url(TestPicture2)",
                      "marginRight": 10,
                    }
                  }
                />
              </div>
            </div>
            <GlyphButton
              className="lmc-collapse-button"
              glyph="chevron-down"
              glyphStyle={Object {}}
              position="default"
              style={
                Object {
                  "paddingBottom": 10,
                }
              }
            />
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcHandoverTitleBar
    createdBy={
      Object {
        "name": Object {
          "first": "Test",
          "last": "Carer",
        },
        "picture": "TestPicture",
      }
    }
    createdOn="Tue Jan 01 2019 00:00:00 GMT+0000 (GMT)"
    onClick={[MockFunction]}
    witnessedBy={
      Object {
        "name": Object {
          "first": "Test2",
          "last": "Carer2",
        },
        "picture": "TestPicture2",
      }
    }
  />,
}
`;
